### ISSUE TITLE
podman build tags images in the local registry with a leading localhost/ and docker does not

### ISSUE THREAD

**Issue by bowlofeggs** [containers/buildah#1034]
**Description**

Does ```podman build``` have the goal of being a drop-in replacement for docker? If not, feel free to close this issue. If so, the issue is that ```podman build -t bodhi-ci/f27``` tags the resulting image into the local engine with a leading ```localhost/``` in the repository name, and ```docker``` does not do this. This results in a difference in how the image must be referenced later when running ```docker/podman run```, because ```docker run… bodhi-ci/f27``` will work, but ```podman run… bodhi-ci/f27``` will not work (and will try to find that image in a variety of external registries).

**Steps to reproduce the issue:**
1. Build a ```Dockerfile``` with ```podman``` and ```docker```, and use the ```-t``` flag.
2. For example, I've built my Bodhi CI images with ```-t bodhi-ci/f27``` in the examples shown below.

**Describe the results you received:**

```
$ sudo docker images
REPOSITORY                          TAG                 IMAGE ID            CREATED             SIZE
bodhi-ci/f27                        latest              4668aacb1a2a        11 hours ago        1.29 GB
$ sudo podman images
REPOSITORY                          TAG      IMAGE ID       CREATED          SIZE
localhost/bodhi-ci/f27              latest   351ffa92bc74   2 minutes ago    1.34GB
```

**Describe the results you expected:**

```
$ sudo podman images
REPOSITORY                          TAG      IMAGE ID       CREATED          SIZE
bodhi-ci/f27              latest   351ffa92bc74   2 minutes ago    1.34GB
```

**Output of `rpm -q buildah` or `apt list buildah`:**

Buildah is not installed, but the text at https://github.com/containers/libpod/issues/new said to file issues with ```podman build``` here.

**Output of `buildah version`:**

See above.

**Output of `cat /etc/*release`:**

```
$ cat /etc/redhat-release 
Fedora release 30 (Rawhide)
```

**Output of `uname -a`:**

```
Linux host.example.com 4.19.0-0.rc3.git3.1.fc30.x86_64 #1 SMP Fri Sep 14 18:31:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Output of `cat /etc/containers/storage.conf`:**

```
# This file is is the configuration file for all tools
# that use the containers/storage library.
# See man 5 containers-storage.conf for more information
# The "container storage" table contains all of the server options.
[storage]

# Default Storage Driver
driver = "overlay"

# Temporary storage location
runroot = "/var/run/containers/storage"

# Primary Read/Write location of container storage
graphroot = "/var/lib/containers/storage"

[storage.options]
# Storage options to be passed to underlying storage drivers

# AdditionalImageStores is used to pass paths to additional Read/Only image stores
# Must be comma separated list.
additionalimagestores = [
]

# Size is used to set a maximum size of the container image.  Only supported by
# certain container storage drivers.
size = ""

# Path to an helper program to use for mounting the file system instead of mounting it
# directly.
#mount_program = "/usr/bin/fuse-overlayfs"

# OverrideKernelCheck tells the driver to ignore kernel checks based on kernel version
override_kernel_check = "true"

# mountopt specifies comma separated list of extra mount options
mountopt = "nodev"

# Remap-UIDs/GIDs is the mapping from UIDs/GIDs as they should appear inside of
# a container, to UIDs/GIDs as they should appear outside of the container, and
# the length of the range of UIDs/GIDs.  Additional mapped sets can be listed
# and will be heeded by libraries, but there are limits to the number of
# mappings which the kernel will allow when you later attempt to run a
# container.
#
# remap-uids = 0:1668442479:65536
# remap-gids = 0:1668442479:65536

# Remap-User/Group is a name which can be used to look up one or more UID/GID
# ranges in the /etc/subuid or /etc/subgid file.  Mappings are set up starting
# with an in-container ID of 0 and the a host-level ID taken from the lowest
# range that matches the specified name, and using the length of that range.
# Additional ranges are then assigned, using the ranges which specify the
# lowest host-level IDs first, to the lowest not-yet-mapped container-level ID,
# until all of the entries have been used for maps.
#
# remap-user = "storage"
# remap-group = "storage"

[storage.options.thinpool]
# Storage Options for thinpool

# autoextend_percent determines the amount by which pool needs to be
# grown. This is specified in terms of % of pool size. So a value of 20 means
# that when threshold is hit, pool will be grown by 20% of existing
# pool size.
# autoextend_percent = "20"

# autoextend_threshold determines the pool extension threshold in terms
# of percentage of pool size. For example, if threshold is 60, that means when
# pool is 60% full, threshold has been hit.
# autoextend_threshold = "80"

# basesize specifies the size to use when creating the base device, which
# limits the size of images and containers.
# basesize = "10G"

# blocksize specifies a custom blocksize to use for the thin pool.
# blocksize="64k"

# directlvm_device specifies a custom block storage device to use for the
# thin pool. Required if you setup devicemapper.
# directlvm_device = ""

# directlvm_device_force wipes device even if device already has a filesystem.
# directlvm_device_force = "True"

# fs specifies the filesystem type to use for the base device.
# fs="xfs"

# log_level sets the log level of devicemapper.
# 0: LogLevelSuppress 0 (Default)
# 2: LogLevelFatal
# 3: LogLevelErr
# 4: LogLevelWarn
# 5: LogLevelNotice
# 6: LogLevelInfo
# 7: LogLevelDebug
# log_level = "7"

# min_free_space specifies the min free space percent in a thin pool require for
# new device creation to succeed. Valid values are from 0% - 99%.
# Value 0% disables
# min_free_space = "10%"

# mkfsarg specifies extra mkfs arguments to be used when creating the base.
# device.
# mkfsarg = ""

# use_deferred_removal marks devicemapper block device for deferred removal.
# If the thinpool is in use when the driver attempts to remove it, the driver 
# tells the kernel to remove it as soon as possible. Note this does not free
# up the disk space, use deferred deletion to fully remove the thinpool.
# use_deferred_removal = "True"

# use_deferred_deletion marks thinpool device for deferred deletion.
# If the device is busy when the driver attempts to delete it, the driver
# will attempt to delete device every 30 seconds until successful.
# If the program using the driver exits, the driver will continue attempting
# to cleanup the next time the driver is used. Deferred deletion permanently
# deletes the device and all data stored in device will be lost.
# use_deferred_deletion = "True"

# xfs_nospace_max_retries specifies the maximum number of retries XFS should
# attempt to complete IO when ENOSPC (no space) error is returned by
# underlying storage device.
# xfs_nospace_max_retries = "0"

# If specified, use OSTree to deduplicate files with the overlay backend
ostree_repo = ""

# Set to skip a PRIVATE bind mount on the storage home directory.  Only supported by
# certain container storage drivers
skip_mount_home = "false"
```

**Comment by TomSweeneyRedHat** [comment]
Thanks for the problem report @bowlofeggs .  Thanks for the thorough problem report, can I bother you to reply with the output of `podman version` please?

podman/buildah build/bud should be a drop in replacement for docker and we'll take a look at this.  However, I thought this had bubbled up before, @nalind or @rhatdan do you recall?  I didn't find it in the issues, but may have missed it.

**Comment by bowlofeggs** [comment]
Sure thing!

```
$ podman --version
podman version 0.9.3-dev
$ rpm -q podman
podman-0.9.3-2.dev.gitc3a0874.fc30.x86_64
```

**Comment by rhatdan** [comment]
Well it is intended to be a drop in replacement for the most part.  The localhost versus docker.io was intentional.

The registry component of an image should indicate where it came from.  And we don't want to hard code the default location of docker.io.

Now if podman can not instantly run the image created by buildah or podman build, then that is a bug.

**Comment by rhatdan** [comment]
It seems like podman run is not looking into localhost if the name containers a repo and a name

@baude could you take a look.

```
podman build -t -t bodhi-ci/f27  ~
podman run -ti bodhi-ci/f27 /bin/sh
```
Fails.
```
podman build -t -t bodhi-ci  ~
podman run -ti bodhi-ci /bin/sh
```
Works.

Our image name processing is failing if the repo is specified.

**Comment by baude** [comment]
For me using master, neither work ... ill look at it sure.

**Comment by rhatdan** [comment]
I believe this is fixed in the latest release.

**Comment by adelton** [comment]
I'm hitting this issue with `podman-0.11.2-15.dev.git75b19ca.fc30.x86_64` installed on Fedora 28 machine. Is the fix supposed to be in Fedora by now?

**Comment by rhatdan** [comment]
It should be in Master branch, and we are working on a podman-0.12.1 right now.

**Comment by jschewebbn** [comment]
I'm using podman 1.6.2 and still seeing this issue of images being prefixed with "localhost/".  Should this still be the case? This makes compatibility with docker scripts difficult as the image names aren't the same.

**Comment by TomSweeneyRedHat** [comment]
I'm still seeing it on v1.8.0 too, but thought it had been fixed.  I'm going to reopen this so it can be chased down.

**Comment by jschewebbn** [comment]
@TomSweeneyRedHat  do you know of a work around that I can use until this is fixed? Is there a way to tell podman to not prefix the tag with "localhost/"?

**Comment by TomSweeneyRedHat** [comment]
I tried playing with it, but I didn't find a quick clean way to do so.  However, when referencing the image, you can ignore the 'localhost/' part of it.  For instance, in the above example you could do:

`podman inspect bodhi-ci/f27`

and that works, does that make it better for your use case @jschewebbn ?

**Comment by jschewebbn** [comment]
The problem that I have is that I'm creating images and then saving them to a docker archive file to transfer and the names of the images have "localhost/" prefixed and then when I try and load the images from the file on the other end with a script not expecting the "localhost/" the script fails.

**Comment by TomSweeneyRedHat** [comment]
@jschewebbn can you tell me the commands that you're using to save the docker archive file and then the command(s) the script is not happy with when trying to load it?

I tried playing with buildah push/pull to a docker-archive and I wasn't seeing the localhost in that output or the inspects that I did.  These are the play commands that I used:
```
buildah push tom docker-archive:/tmp/tom:latest
buildah pull docker-archive:/tmp/tom
```

**Comment by jschewebbn** [comment]
```

**Comment by jschewebbn** [comment]
The workflow that I've been doing with docker is this:
```
cd image1
docker build -t image1 .
cd ../image2
docker build -t image2 .
... # create all of the images

rm -f archive.tar
docker save -o archive.tar image1 image2 ...
```
Copy archive.tar to another host.
```
docker load -i archive.tar
docker pull registry:2
docker run ... registry:2

for image in image1 image2 ...; do
    docker tag ${image} ${host}:5000/v2/${image}
    docker push ${host}:5000/v2/${image}
done
```

Perhaps I'm doing some things incorrectly as my docker knowledge is pretty basic.

**Comment by sttyerase** [comment]
```
Fedora Rules>>podman images dbmidocker
REPOSITORY                  TAG   IMAGE ID       CREATED         SIZE
localhost/dbmidocker/hfb0   1.0   0378f49b278e   8 minutes ago   638 MB
Fedora Rules>>podman version
Version:            1.8.1
RemoteAPI Version:  1
Go Version:         go1.13.6
OS/Arch:            linux/amd64
Fedora Rules>>cat /etc/redhat-release
Fedora release 31 (Thirty One)
Fedora Rules>>

```
Same issue here.

**Comment by rhatdan** [comment]
@vrothberg PTAL, I think this is around the saving with multiple images?

**Comment by vrothberg** [comment]
> @vrothberg PTAL, I think this is around the saving with multiple images?

It's a different issue although it affects save and load as well. Podman and Buildah is prefexing short images (or not fully-qualified images without a domain/registry) with "localhost/". That's inconsistent with Docker's behaviour but consistent among Podman and Buildah.

Personally, I don't like the "localhost/" because the tools don't what I want them to but I also see wisdom explicitly enforcing fully-qualified image names.

I am closing the issue as changing the behaviour will very likely break users who already rely in the "localhost/" prefixes. However, in some cases it might cause issues migrating from Docker to Podman/Buildah.

**Comment by jschewebbn** [comment]
I take this closing of the ticket to mean that podman does not want to be compatible with docker. This will make it very difficult for those of us advocating people move from docker to podman as they now have an incompatibility that there is no work around for. If podman were to have a command line option to suppress this "localhost/" prefix that would be sufficient. Then I could tell people that podman save works just like docker, except you need this extra flag and it works.

**Comment by vrothberg** [comment]
@jschewebbn, thanks for continuing the conversation!

Did you encounter a migration issue regarding this behaviour?

Podman will treat "foo" and "localhost/foo" as the same image, so scripts that operated on Docker can safely migrate over to Podman as "foo" will implicitly resolve to "localhost/foo". However, `podman images` will list "localhost/foo" which is the only incompatibility I know.

**Comment by jschewebbn** [comment]
I don't know what a migration issue is. I've been following this issue as this is the problem that I'm having. In my specific use case I would like to create images with podman and then load them with docker. See https://github.com/containers/buildah/issues/1034#issuecomment-587562980 for the details.

**Comment by rhatdan** [comment]
So you want podman save of foo to save it as docker.io/library/foo?

**Comment by jschewebbn** [comment]
I want `podman save foo` to create a file that has the same `manifest.json` and `repositories` files that `docker save foo` has. 

For example when I execute `docker save foo` I see this in `manifest.json`:
```json
[
  {
    "Config": "cd998886946a19daf9b43f35476427b9e800999f343ca463eae89f4570c10cc9.json",
    "RepoTags": [
      "foo:latest"
    ],
    "Layers": [
...
    ]
  }
]
```

When I executed `podman save foo` I see this in `manifest.json`:
```json
[
  {
    "Config": "cd998886946a19daf9b43f35476427b9e800999f343ca463eae89f4570c10cc9.json",
    "RepoTags": [
      "localhost/foo:latest"
    ],
    "Layers": [
...
    ]
  }
]
```

The key here is the prefix on the "RepoTags" property.

The `repositories` file has a similar issue:
docker shows
```json
{
  "foo": {
    "latest": "cdd290f400b43594611786fe015d8f78205421db0bd8afbbf3a12d55fffdd1ed"
  }
}
```

podman shows
```json
{
  "localhost/frc":{
      "latest":"92b78c85f2b4b16c8ea4f6235af775f09d227062270fdd84bd7052c19d850528"
    }
}
```

**Comment by ssbarnea** [comment]
Any plans to address this? This is a deal breaker for allowing both engines to be used in parallel even for the most basic CI/CD tasks, like:

```bash
ENGINE=`shell command -v podman docker|head -n1`
$ENGINE build -t elastic-recheck .
# docker only:
$ENGINE run -it elastic-recheck /bin/bash
# podman only:
$ENGINE run -it localhost/elastic-recheck /bin/bash
```

**Comment by vrothberg** [comment]
@rhatdan, as you reopened. What's your take on it?

The localhost prefixing is incompatible with Docker. We've been discussing this issue a couple of times already but the conclusion was to keep the behaviour. I am no fan of it because if I tag something as `foo` I want it to be `foo`.

However, if we changed the behaviour, it would be a breaking change for both, Podman and Buildah.

**Comment by vrothberg** [comment]
> Any plans to address this? This is a deal breaker for allowing both engines to be used in parallel even for the most basic CI/CD tasks, like:
> 
> ```shell
> ENGINE=`shell command -v podman docker|head -n1`
> $ENGINE build -t elastic-recheck .
> # docker only:
> $ENGINE run -it elastic-recheck /bin/bash
> # podman only:
> $ENGINE run -it localhost/elastic-recheck /bin/bash
> ```

Note that it's only breaking when enforcing exactly this scenario. Doing a `build -t localhost/...` or ideally using a fully-qualified image reference will work across Docker, Buildah and Podman.

**Comment by jschewebbn** [comment]
If you want to avoid breaking existing Podman behavior, a command line switch that would be acceptable. It would make the ENGINE example above a bit more complicated, one would need to check which engine is being used and possibly add additional arguments, but that's not very difficult.

The really slow and inefficient workaround is to write a script that fixes the image after it's been built.

**Comment by ssbarnea** [comment]
Thanks for the hints, I was able to add the localhost/ prefix in order to make it work for both.

How about a compromise and raise warning on stderr every time a "- t tag-with-no-prefix" is called from the command line, asking user to mention the prefix in order to avoid portability issues?

**Comment by jschewebbn** [comment]
@ssbarnea  the problem with just warning is that doesn't get the compatibility with docker that some of us need. I need to have a way to tell podman to not include a prefix. It's fine to get a warning in this case, but there needs to be a way to say no prefix.

**Comment by jschewebbn** [comment]
Here's a python script that one can use in place of `docker save` that handles multiple images and removes the "localhost" prefix that podman adds. While this works, it's really inefficient. However until podman adds an option to not specify the prefix, this is what I need to do.
[podman_save_multiple.txt](https://github.com/containers/buildah/files/5210158/podman_save_multiple.txt)

**Comment by TomSweeneyRedHat** [comment]
This sounds like an excellent weekend project for myself.  Let me see if I can come up with a way to use a new option to remove localhost if the option is set to true.  `--nolocalhost` perhaps?

**Comment by vrothberg** [comment]
> This sounds like an excellent weekend project for myself. Let me see if I can come up with a way to use a new option to remove localhost if the option is set to true. `--nolocalhost` perhaps?

I suggest to wait until we have consensus on the way forward.

IMHO a flag isn't as user-friendly as I'd like it to be. If users wish to change the behaviour, they had to specify the flag all the time which is a bit annoying and error prone as it can easily be forgotten.

I could imagine, however, a new option in the `registries.conf`.

**Comment by rhatdan** [comment]
Why is this Docker only?
```
# docker only:
$ENGINE run -it elastic-recheck /bin/bash
```
podman run -it elastic-recheck /bin/bash 

Works on Podman?

**Comment by rhatdan** [comment]
All podman commands are supposed to prepend the localhost/ prefix on images without a registry.  Just like Docker does with docker.io.  Am I missing something?
Only issue I see is the podman images, which we could make hide the localhost/ prefix if you would want?

**Comment by jschewebbn** [comment]
I want to create an image with podman named "foo", then save it to a file and then load the image with docker and the image to be named "foo". This is consistent with how docker works. If I need to add a command line flag to the podman save command, that's fine. If I need to put some flag in my podman registries.conf, that's fine too. 

Currently I'm running the python script that I posted in https://github.com/containers/buildah/issues/1034#issuecomment-691175765 to export my images from podman. This makes podman compatible with docker at the expense of needing to rewrite the image files and export the same layers multiple times. The benefit here is that I don't need to run as root to create my images. If it weren't for that, I'd just use docker.

**Comment by vrothberg** [comment]
> All podman commands are supposed to prepend the localhost/ prefix on images without a registry. Just like Docker does with docker.io. Am I missing something?

Docker only normalizes to docker.io when pushing (or pulling). Podman prefixes localhost/ when _tagging_. So when you run `podman tag foo` and want to check with `podman images` you don't get foo but localhost/foo.

Podman push and pull behave differently. AFAIR push will pick the first item in the search registries and pull iterates over the list with first successful pull wins.

**Comment by pclass-sensonix** [comment]
I'd be interested in being able to remove the localhost/ prefix for the image names with podman as well. As another use case, I think it is preventing me from remotely developing on a Linux box running podman from PyCharm on my Mac. 

<img width="615" alt="Screen Shot 2020-12-04 at 5 28 09 PM" src="https://user-images.githubusercontent.com/61421504/101225223-83210a00-3656-11eb-92e4-875176e5081e.png">
<img width="841" alt="Screen Shot 2020-12-04 at 5 22 10 PM" src="https://user-images.githubusercontent.com/61421504/101225224-84523700-3656-11eb-9b7d-f1d2a660f1aa.png">

**Comment by rhatdan** [comment]
@vrothberg Let's talk about this next week, perhaps at Watercooler.

**Comment by rhatdan** [comment]
@vrothberg we never went over this.  Perhaps we could use this use case.  Would hiding localhost from users be helpful?

**Comment by vrothberg** [comment]
> @vrothberg we never went over this. Perhaps we could use this use case. Would hiding localhost from users be helpful?

I don't think that's helpful as it would cause confusion on the other side when doing a `podman tag $image localhost/foo; podman images`.

However, Podman treats "foo" and "localhost/foo" in the very same way.  We unified the code across Buildah, CRI-O and POdman with the short-names work.

**Comment by pclass-sensonix** [comment]
Is this short-names work available so I could try it out? Thanks.

**Comment by vrothberg** [comment]
Yes, I recommend reading the following blog to get started: https://www.redhat.com/sysadmin/container-image-short-names

**Comment by VictorCallejas** [comment]
Hi everyone,

I am having the same issue. I testing a kubernetes cluster with docker and cri-o runtimes.

When loading the same tar images with docker and podman I get different repository names which produces errors:
`Failed to pull image "calico/cni:v3.16.0"`

Podman images:
```
REPOSITORY                                        TAG         IMAGE ID      CREATED        SIZE
localhost/calico/cni                              v3.16.0     1a7fba483100  6 months ago   133 MB
```

Docker images:
```
REPOSITORY                                        TAG         IMAGE ID      CREATED        SIZE
calico/cni                                                     v3.16.0             1a7fba483100        6 months ago        133MB
```
This only happens with some images.

Do you know any way to workaround this?

Thank you.

**Comment by rhatdan** [comment]
If I pull the image from docker.io

```
$ podman pull calico/cni   
...
$ podman images | grep calico
docker.io/calico/cni                 v3.16.0  1a7fba483100  6 months ago  133 MB
```

In Docker, docker.io is assumed, so the image looks like:
```
docker images | grep calico
calico/cni                                    latest                 21fdaa2fccee   12 days ago     131MB
```

If you build the images locally  then podman sets the "registry" to localhost, while docker sets it to docker.io.

**Comment by VictorCallejas** [comment]
@rhatdan thank you very much for your fast response.

But that's not possible because it is a closed cluster without access to the internet.

I need to load images from local tar files.

**Comment by rhatdan** [comment]
Ok, can't you tag the images?

**Comment by jschewebbn** [comment]
Yes, we can tag images, but they will still have the "localhost:" prefix. We can't control if that is in the name or not. See https://github.com/containers/buildah/issues/1034#issuecomment-692924006 for the use case and a script that allows one to have this control.
I understand that this doesn't follow the philosophy of podman in that you can't tell which registry the image came from. However we're looking for docker compatibility. This is an area where podman is not compatible with docker and without using the referenced script cannot be made compatible with docker.

**Comment by rhatdan** [comment]
How about using podman import/export rather then save/load?

**Comment by jschewebbn** [comment]
Can you do this?

**Comment by rhatdan** [comment]
Does this work for you?

```
# podman build -t myimage -f /tmp/test/Dockerfile /tmp/test/
STEP 1: FROM alpine
STEP 2: COMMIT myimage
--> 28f6e270574
28f6e27057430ed2a40dbdd50d2736a3f0a295924016e294938110eeb8439818
# podman images | grep myimage
localhost/myimage          latest    28f6e2705743  4 weeks ago  5.88 MB
# podman tag myimage docker.io/myimage
# podman images | grep myimage
localhost/myimage          latest    28f6e2705743  4 weeks ago  5.88 MB
docker.io/library/myimage  latest    28f6e2705743  4 weeks ago  5.88 MB
# podman save docker.io/library/myimage -o /tmp/myimage.tar
# docker load -i /tmp/myimage.tar 
cb381a32b229: Loading layer [==================================================>]   5.88MB/5.88MB
Loaded image: myimage:latest
# docker images | grep myimage
myimage      latest    28f6e2705743   4 weeks ago   5.61MB
```

**Comment by jschewebbn** [comment]
No, that doesn't work.
```
>podman --version
podman version 3.0.1

>podman image ls | grep flink_front
localhost/flink_frontend                         latest  306c4ca6b7b2  5 weeks ago   705 MB

>podman tag flink_frontend flink_frontend
>podman save -o test flink_frontend
>docker load -i test

>docker image ls -a
REPOSITORY                 TAG                 IMAGE ID            CREATED             SIZE
localhost/flink_frontend   latest              306c4ca6b7b2        5 weeks ago         695MB

```

"localhost/" should not be there to be compatible with docker save/load.

I looked at `export` and I see that wants to reference a container. I don't have a container, I just have a filesystem image.

If the text on the podman main page "Simply put: alias docker=podman" is going to work for people, there needs to be a way to remove this prefix. Maybe the alias becomes "alias docker='podman --docker-compatibility'" and then things like this could be handled.

**Comment by rhatdan** [comment]
You didn't do what I suggested.

podman tag flink_frontend docker.io/flink_frontend
podman save -o test docker.io/flink_frontend

**Comment by jschewebbn** [comment]
OK, that worked. Now why did that work? Is "docker.io" special in some way?

**Comment by rhatdan** [comment]
Yes Docker hard codes the assumption of all images come from docker.io if a registry is not specified.  Where as Podman hard codes localhost to indicate no registry is involved.

**Comment by rhatdan** [comment]
That is part of their business mode to lock you into docker.io.

**Comment by jschewebbn** [comment]
Except that when I do all of this with docker the local registry is used and there is no prefix. While this will work, it doesn't fit into the "drop in replacement" model for docker. Thank you for the work around.

**Comment by VictorCallejas** [comment]
Hi everyone, thanks for all the advices.

My use case for this problem was to test docker, containerd and cri-o runtimes in the same cluster without changing any kubernetes references.

When you save images to tar files with docker, sometimes docker.io is overlooked as it can be infer as default, this does not happen with podman.

The problem is that **docker assumes docker.io as default registry** and podman not, **podman prepends 'localhost/' when there is no registry**.

So when you load the images you have this:

Podman images:
```
REPOSITORY                                        TAG         IMAGE ID      CREATED        SIZE
localhost/calico/cni                              v3.16.0     1a7fba483100  6 months ago   133 MB
```

Docker images:
```
REPOSITORY                                        TAG         IMAGE ID      CREATED        SIZE
calico/cni                                                     v3.16.0             1a7fba483100        6 months ago        133MB
```

This way you need to change image references.

For my use case, the easiest solution was to save all images with podman instead of docker, and then depending on runtime load them with docker or podman, but **this way docker.io will always be explicit** and there is no need to infer anything (docker.io or localhost).

Hope it helps.
Cheers.

**Comment by dajtxx** [comment]
I'm trying to move from docker to podman, and tried to bring up a docker-compose stack by pointing docker-compose to the podman unix socket.

I have to change the image names of my locally built images in my docker-compose.yml file to have the localhost/ prefix or it doesn't work. The postgres and rabbitmq images are fine as-is.

So this is another case where podman is not a drop-in replacement for docker.

Perhaps working with docker-compose is outside the scope of the podman project. I was just looking for the simplest possible way to get my project moved over while I learn about podman.

**Comment by dajtxx** [comment]
The images from docker.io worked fine, it was the locally built ones that podman puts `localhost/` in front of that are the 'problem'.

It looks like this was a design decision by the podman team, so it's not a bug. I put my comment in because google brought me to this issue and I thought it might be useful for others to see another option for getting around that when google brings them here.

**Comment by rhatdan** [comment]
Right locally built images do not default to any registry. Perhaps a modification to this would be to allow users to specify a default registry, to eliminate this problem.

**Comment by TonioGela** [comment]
So 5 years later, the interoperability with Docker in scripts and|or third party applications still isn't a thing?

**Comment by rhatdan** [comment]
If you want us to hard code docker.io prefix, then no it is not.

**Comment by nitind** [comment]
Must there be a prefix at all?

**Comment by rhatdan** [comment]
I don't believe we can support ""/imagename

**Comment by robrich** [comment]
+1 for removing the `localhost/` prefix. How does this pain me? Kind doesn't assume that `localhost/myimage:v0.1` and `myimage:v0.1` are the same. So now I need to modify all my k8s yaml files to add the `localhost/` prefix.

**Comment by rhatdan** [comment]
What are you running your k8s yaml files with?  podman kube play?

**Comment by adelton** [comment]
I for one have changed all my CI's and habits to explicitly say `localhost/` whenever I mean an image not coming from remote registry, like locally built. It adds the prefix for docker operations, and it does not break podman. It is kind of nice to be explicit.

**Comment by sttyerase** [comment]
Explicit is nice in one sense.  It makes intentions clear to others.  That's why I like the POSIX --someexplicitoptionname convention over -s.  However, if you want Podman to be interchangeable with Docker, it's a show stopper.  If you want to force individuals, groups and companies to choose between Docker and Podman, then keep the localhost prefix.

This issue is so old, I don't know why it came up now.  I long  ago abandoned it and forgot I had a watch on it.

**Comment by robrich** [comment]
> What are you running your k8s yaml files with? podman kube play?

I'm using Podman to build images, Podman Desktop to get the images to the kind cluster, and kubectl to start the containers given a yaml file.

**Comment by rhatdan** [comment]
Docker defaulted the locally built image to docker.io, hard coding their business use case into their tooling.  Podman will not do the same.

Images should be pushed to a container registry, at which point they will be named with the registry.  localhost/image will become docker.io/image if pushed to docker.io or quay.io/image if pushed to quay.io.

**Comment by jschewebbn** [comment]
Actually the locally built docker images don't have any prefix associated with them. With podman they have "localhost/". The workflow that is a problem is exporting the locally created images and sharing them as they have different names when created in podman than when created in docker.

**Comment by vrothberg** [comment]
Docker will always normalize short names to docker.io.  So an image `foo` will be normalized to `docker.io/library/foo:latest`. Podman decided to normalize short names (when tagging!) to `localhost/foo:latest`.  This behavior is so old that changing it would break existing workloads.

I can imagine adding an opt-in mechanism to avoid the `localhost/` prefix.  This may impact other functionality though and hence requires some more thought.  But let's not decide that in a 5 years old issue that has been closed years ago.  If you have a strong desire for such an opt-in switch, I want to encourage you to open a fresh issue.

**Comment by rhatdan** [comment]
If you have an example of a docker-compose script that blows up in this manner please open a new issue.

Podman should be able to compensate and find an image based on shortname if built locally.  BTW Is this happening on a MAC or Windows Box?

**Comment by mikhey** [comment]
What about this as a possible solution? Instead of removing the `localhost/` tag prefix when referencing images podman searches for tags prefixed with `localhost/` first? 

So if my DockerFile is referencing `bodhi-ci/f27` it will search for an image tagged `localhost/bodhi-ci/f27`?  

It could be setting possibly to tell podman to use it's own list of image `localhost/` prefixed images *first* before attempting to look for it at repository X's URL? 

It makes sense to me why the repo name tag prefix is visible - that provides clarity as to the origin of the image.  

----

I came across this issue when attempting to use [C# (.NET) and MS SQL (dotnet-mssql)](https://github.com/devcontainers/templates/tree/main/src/dotnet-mssql).  During the build process it creates a temporary image tag and expects to be able to use it w/o the `localhost/` prefix resulting in an error.  

I'm looking for a way around it by possibly manually building the images that devcontainer command would and modifying the docker-compose.yml file, but if podman was aware of it's `localhost/` tagged images, it would make life a little easier/more compatible.

**Comment by rhatdan** [comment]
Please open new issues or discussions.
